src/System.Data.SqlClient/src/System/Data/SqlClient/SqlSequentialTextReader.cs

```diff
@@ -391,20 +379,15 @@ private byte[] PrepareByteBuffer(int numberOfChars, out int byteBufferUsed)
         /// <param name="inBuffer">Buffer of bytes to decode</param>
         /// <param name="inBufferCount">Number of bytes to decode from the inBuffer</param>
         /// <param name="outBuffer">Buffer to write the characters to</param>
-        /// <param name="outBufferOffset">Offset to start writing to outBuffer at</param>
-        /// <param name="outBufferCount">Maximum number of characters to decode</param>
         /// <returns>The actual number of characters decoded</returns>
-        private int DecodeBytesToChars(byte[] inBuffer, int inBufferCount, char[] outBuffer, int outBufferOffset, int outBufferCount)
+        private int DecodeBytesToChars(byte[] inBuffer, int inBufferCount, Span<char> outBuffer)
         {
             Debug.Assert(inBuffer != null, "Null input buffer");
             Debug.Assert((inBufferCount > 0) && (inBufferCount <= inBuffer.Length), $"Bad inBufferCount: {inBufferCount}");
             Debug.Assert(outBuffer != null, "Null output buffer");
-            Debug.Assert((outBufferOffset >= 0) && (outBufferCount > 0) && (outBufferOffset + outBufferCount <= outBuffer.Length), $"Bad outBufferCount: {outBufferCount} or outBufferOffset: {outBufferOffset}");
+            Debug.Assert((outBuffer.Length > 0), $"Bad outBuffer size: {outBuffer.Length}");
 
-            int charsRead;
-            int bytesUsed;
-            bool completed;
-            _decoder.Convert(inBuffer, 0, inBufferCount, outBuffer, outBufferOffset, outBufferCount, false, out bytesUsed, out charsRead, out completed);
+            _decoder.Convert(inBuffer.AsSpan(..inBufferCount), outBuffer, false, out int bytesUsed, out int charsRead, out bool completed);
```

Comment:
How are you validating that so if these changes are a measurable improvement?

Reply:
No measurements made, but apisof.net suggests TextReader https://apisof.net/catalog/System.IO.TextReader.Read() has quite some widespread use, likely in a very hot path. So allocating a char[] for each char read is highly likely going to yield a measurable improvement.
I would probably even go further and override the Span based reads of TextReader.

Reply:
I’d argue if the performance isn’t regressed and the heap allocation is removed and the code isn’t more complicated then it’s a no brainer.

My 2c

Reply:
> but apisof.net suggests TextReader https://apisof.net/catalog/System.IO.TextReader.Read() has quite some widespread use

TextReader, yes, but that's mainly StreamReader.  In order to get one of these SqlSequentialTextReaders, you call SqlClient.GetTextReader, and that has effectively zero usage in apisof.net's data:
https://apisof.net/catalog/System.Data.SqlClient.SqlDataReader.GetTextReader(Int32)

> I’d argue if the performance isn’t regressed and the heap allocation is removed and the code isn’t more complicated then it’s a no brainer.

If there's no measurable improvement, what's the benefit of the churn?  We unfortunately have poor testing of this area; any change has a reasonable shot at regressing something.

Reply:
I agree with @stephentoub 
The SqlClient testing happens offline and is a slow process right now because we test a bunch of libraries which use SqlClient. Please decouple the SqlClient changes from this PR and send them out as a separate PR. 
Also some proof of perf improvements would be nice.

Reply:
I reverted the changes for this PR. I will see if I can make some measurements and make a separate PR if I deem the improvements to be worthwhile!